### PR TITLE
:seedling: syncer(apiimporter,syncer,tunneler}: use structured, contextual logging

### PR DESCRIPTION
**Pull request by dinhxuanvu**
Signed-off-by: Vu Dinh <vudinh@outlook.com>

## Related issue(s)

Part of https://github.com/kcp-dev/kcp/issues/1694

### CODE REVIEW

**Comment by dinhxuanvu** [comment]
@stevekuznetsov FYI. I hope I did it right.

File: pkg/syncer/apiimporter.go [hunk 1]
```diff
@@ -136,42 +136,47 @@ func (i *APIImporter) Start(ctx context.Context, pollInterval time.Duration) {
 
 	i.kcpInformerFactory.Start(ctx.Done())
 	i.kcpInformerFactory.WaitForCacheSync(ctx.Done())
-
-	klog.Infof("Starting API Importer for location %s in cluster %s", i.location, i.logicalClusterName)
+	logger := klog.FromContext(ctx)
+	logger.WithValues("location", i.location, "logical-cluster", i.logicalClusterName)
```
Comment (stevekuznetsov) [inline pkg/syncer/apiimporter.go:140]: Could we also add the reconciler key with "api-importer" or something similar?

File: pkg/syncer/apiimporter.go [hunk 2]
```diff
@@ -136,42 +136,47 @@ func (i *APIImporter) Start(ctx context.Context, pollInterval time.Duration) {
...
 	<-ctx.Done()
-	i.Stop()
+	i.Stop(ctx)
 }
 
-func (i *APIImporter) Stop() {
-	klog.Infof("Stopping API Importer for location %s in cluster %s", i.location, i.logicalClusterName)
+func (i *APIImporter) Stop(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+	logger.WithValues("location", i.location, "logical-cluster", i.logicalClusterName)
```
Comment (stevekuznetsov) [inline pkg/syncer/apiimporter.go:154]: Whenever you find yourself adding the same fields to the logger more than once, consider if it's possible to add them to a common context at a higher level and passing the logger via this single context down to all callers.

File: pkg/syncer/apiimporter.go [hunk 3]
```diff
@@ -189,22 +194,22 @@ func (i *APIImporter) ImportAPIs(ctx context.Context) {
 			clusterctl.GetGVRForLocationInLogicalClusterIndexKey(i.location, i.logicalClusterName, gvr),
 		)
 		if err != nil {
```
Comment (stevekuznetsov) [inline pkg/syncer/apiimporter.go:196]: nit: add the group+resource to the logger in the loop, so each logging call for a particular GR has the data

File: pkg/syncer/apiimporter.go [hunk 4]
```diff
@@ -189,22 +194,22 @@ func (i *APIImporter) ImportAPIs(ctx context.Context) {
...
-			klog.Errorf("error pulling CRDs: %v", err)
+			logger.Error(fmt.Errorf("error pulling CRDs: %v", err))
```
Comment (stevekuznetsov) [inline pkg/syncer/apiimporter.go:197]:
```suggestion
			logger.Error(err, "error pulling CRDs")
```

File: pkg/syncer/apiimporter.go [hunk 5]
```diff
@@ -189,22 +194,22 @@ func (i *APIImporter) ImportAPIs(ctx context.Context) {
...
 			continue
 		}
 		if len(objs) > 1 {
-			klog.Errorf("There should be only one APIResourceImport of GVR %s for location %s in logical cluster %s, but there was %d", gvr.String(), i.location, i.logicalClusterName, len(objs))
+			logger.Error(fmt.Errorf("There should be only one APIResourceImport of GVR %s but there was %d", gvr.String(), len(objs)))
```
Comment (stevekuznetsov) [inline pkg/syncer/apiimporter.go:201]:
```suggestion
			logger.Error(fmt.Errorf("there should be only one APIResourceImport of GVR %s but there was %d", gvr.String(), len(objs)))
```

File: pkg/syncer/apiimporter.go [hunk 6]
```diff
@@ -189,22 +194,22 @@ func (i *APIImporter) ImportAPIs(ctx context.Context) {
...
 			continue
 		}
 		if len(objs) == 1 {
 			apiResourceImport := objs[0].(*apiresourcev1alpha1.APIResourceImport).DeepCopy()
 			if err := apiResourceImport.Spec.SetSchema(crdVersion.Schema.OpenAPIV3Schema); err != nil {
-				klog.Errorf("Error setting schema: %v", err)
+				logger.Error(fmt.Errorf("Error setting schema: %v", err))
```
Comment (stevekuznetsov) [inline pkg/syncer/apiimporter.go:207]:
```suggestion
				logger.Error(err, "error setting schema")
```

Comment (stevekuznetsov) [reply pkg/syncer/apiimporter.go:207]: (won't comment on every one, but there are many like this)

File: pkg/syncer/syncer.go [hunk 7]
```diff
@@ -109,7 +111,7 @@ func StartSyncer(ctx context.Context, cfg *SyncerConfig, numSyncerThreads int, i
 		}
 
 		if len(syncTarget.Status.VirtualWorkspaces) > 1 {
-			klog.Errorf("SyncTarget %s|%s should not have several Syncer virtual workspace URLs: not supported for now, ignoring additional URLs", cfg.SyncTargetWorkspace, cfg.SyncTargetName)
+			logger.Error("SyncTarget should not have several Syncer virtual workspace URLs: not supported for now, ignoring additional URLs")
```
Comment (stevekuznetsov) [inline pkg/syncer/syncer.go:114]: Does this compile?

File: pkg/syncer/syncer.go [hunk 8]
```diff
@@ -169,15 +171,15 @@ func StartSyncer(ctx context.Context, cfg *SyncerConfig, numSyncerThreads int, i
 	// syncers depend on the types being present to start their informers.
 	var gvrs []schema.GroupVersionResource
 	err = wait.PollImmediateInfinite(gvrQueryInterval, func() (bool, error) {
-		klog.Infof("Attempting to retrieve GVRs from upstream...")
+		logger.Info("Attempting to retrieve GVRs from upstream...")
```
Comment (stevekuznetsov) [inline pkg/syncer/syncer.go:174]: nit: try to use lower-case

File: pkg/syncer/syncer.go [hunk 9]
```diff
@@ -190,11 +192,11 @@ func StartSyncer(ctx context.Context, cfg *SyncerConfig, numSyncerThreads int, i
 	// Check whether we're in the Advanced Scheduling feature-gated mode.
 	advancedSchedulingEnabled := false
 	if syncTarget.GetAnnotations()[AdvancedSchedulingFeatureAnnotation] == "true" {
-		klog.Infof("Advanced Scheduling feature is enabled for syncTarget %s", cfg.SyncTargetName)
+		logger.Info("Advanced Scheduling feature is enabled")
 		advancedSchedulingEnabled = true
 	}
 
-	klog.Infof("Creating spec syncer for SyncTarget %s|%s, resources %v", cfg.SyncTargetWorkspace, cfg.SyncTargetName, resources)
+	logger.Info(fmt.Sprintf("Creating spec syncer resources %v", resources))
```
Comment (stevekuznetsov) [inline pkg/syncer/syncer.go:199]: nit: as a rule of thumb, `fmt.Sprintf()` in structured logs is a smell - attach the data you're formatting as fields

**Review by stevekuznetsov — APPROVED** [review]
Thanks @dinhxuanvu !!

**Pull request merged**